docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+properties in the pageserver cluster without implementing a consensus mechanism directly
+in the pageservers.
+
+## Motivation
+
+Currently, the pageserver's remote storage format does not provide a mechanism for addressing
+split brain conditions that may happen when replacing a node or when migrating
+a tenant from one pageserver to another.
+
+From a remote storage perspective, a split brain condition occurs whenever two nodes both think
+they have the same tenant attached, and both can write to S3. This can happen in the case of a
+network partition, pathologically long delays (e.g. suspended VM), or software bugs.
+
+In the current deployment model, control plane guarantees that a tenant is attached to one
+pageserver at a time, thereby ruling out split-brain conditions.
+However, there is always the risk of a control plane bug.
+
+Futher, lack of safety during split-brain conditions blocks two important features where occasional
+split-brain conditions are part of the design assumptions:
+* seamless tenant migration ([RFC PR](https://github.com/neondatabase/neon/pull/5029))
+* automatic pageserver instance failure handling (aka "failover") (RFC TBD)
+
+### Prior art
+
+- 020-pageserver-s3-coordination.md
+- 023-the-state-of-pageserver-tenant-relocation.md
+- 026-pageserver-s3-mvcc.md
+
+This RFC has broad similarities to the proposal to implement a MVCC scheme in
+S3 object names, but this RFC avoids a general purpose transaction scheme in
+favour of more specialized "generations" that work like a transaction ID that
+always has the same lifetime as a pageserver process and/or tenant attachment.
+
+## Requirements
+
+- Accommodate storage backends with no atomic or fencing capability (i.e. work within
+  S3's limitation that there are no atomics and clients can't be fenced)
+- Don't depend on any STONITH or node fencing in the compute layer (i.e. we will not
+  assume that we can reliably kill and EC2 instance and have it die)
+- Scoped per-tenant, not per-pageserver; for *seamless tenant migration*, we need
+  per-tenant granularity, and for *failover*, we likely want to spread the workload
+  of the failed pageserver instance to a number of peers, rather than monolithically
+  moving the entire workload to another machine.
+  We do not rule out the latter case, but should not constrain ourselves to it.
+
+## Design Tenets
+
+These are not requirements, but are ideas that guide the following design:
+
+- Avoid implementing another consensus system: we already have a strongly consistent
+  database in the control plane that can do atomic operations where needed, and we also
+  have a Paxos implementation in the safekeeper.
+- Avoiding locking in to specific models of how failover will work (e.g. do not assume that
+  all the tenants on a pageserver will fail over as a unit).
+- Avoid doing synchronization that scales with the number of tenants, unless absolutely
+  necessary.
```

Comment:
I didn't understand this. Can you elaborate?

If "synchronization" means attach/detach a tenant, we do those operations on a per-tenant basis. So that _does_ scale linearly with the number of tenants.

Reply:
Added "For example, a pageserver should be able to start up without having to have each tenant individually re-attached."

Right now the way the control plane <-> pageserver interaction works is quite heavyweight per-tenant (REST requests, goroutines running `Operation`) -- this will matter less in future as we scale the system and maybe move to something a bit more queue-like for pushing changes down to pageserver.

If we implement totally ephemeral pageserver identities (there's an "Ephemeral node IDs" in the appendix about interop with other features), then we could implement a model where we explicitly re-attach things on startup, but in the process I think we'll have to deal with scaling issues, and I'd like to sidestep that in the short term by just avoiding doing these really big O(tenants) operations.

Reply:
This bullet is gone now -- we have compromised + now do an O(tenants) op during /re-attach